pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -26,301 +30,214 @@
-  private static final int INT_SIZE = Integer.SIZE / Byte.SIZE;
-  // Maximum number of entries for each buffer
-  public static final int MAX_NUM_ENTRIES = Integer.MAX_VALUE / INT_SIZE;
-
-  private final File invertedIndexFile;
-  private final FieldSpec spec;
-  long start = 0;
-  private ByteBuffer origValueBuffer;
-  private ByteBuffer origLengths;
-  IntBuffer valueBuffer;
-  IntBuffer lengths;// used in multi value
-  int currentVacantPos = 0;
-  private int cardinality;
-  private int capacity;
-  private ByteBuffer origPostingListBuffer;
-  private ByteBuffer origPostingListLengths;
-  private ByteBuffer origPostingListStartOffsets;
-  private ByteBuffer origPostingListCurrentOffsets;
-  IntBuffer postingListBuffer; // entire posting list
-  IntBuffer postingListLengths; // length of each posting list
-  IntBuffer postingListStartOffsets; // start offset in posting List Buffer
-  IntBuffer postingListCurrentOffsets; // start offset in posting List Buffer
-  private int numDocs;
-
-  public OffHeapBitmapInvertedIndexCreator(File indexDir, int cardinality, int numDocs, int totalNumberOfEntries,
-      FieldSpec spec) {
-    String columnName = spec.getName();
-    Preconditions.checkArgument((cardinality > 0) && (cardinality <= MAX_NUM_ENTRIES),
-        "For column: %s, cardinality: %s must > 0 and <= %s", columnName, cardinality, MAX_NUM_ENTRIES);
-    Preconditions.checkArgument((numDocs > 0) && (numDocs <= MAX_NUM_ENTRIES),
-        "For column: %s, numDocs: %s must > 0 and <= %s", columnName, numDocs, MAX_NUM_ENTRIES);
-    if (!spec.isSingleValueField()) {
-      Preconditions.checkArgument((totalNumberOfEntries > 0) && (totalNumberOfEntries <= MAX_NUM_ENTRIES),
-          "For column: %s, totalNumberOfEntries: %s must > 0 and <= %s for multi-value column", columnName,
-          totalNumberOfEntries, MAX_NUM_ENTRIES);
+  private static final int INT_SIZE_IN_BYTES = Integer.SIZE / Byte.SIZE;
+
+  // Maximum number of values due to 2GB file size limit
+  public static final int MAX_NUM_VALUES = Integer.MAX_VALUE / INT_SIZE_IN_BYTES;
+
+  private final String _columnName;
+  private final boolean _singleValue;
+  private final int _cardinality;
+  private final int _numDocs;
+  private final int _numValues;
+  private final File _indexDir;
+  private final File _invertedIndexFile;
+
+  private ByteBuffer _valueByteBuffer;
+  private IntBuffer _valueIntBuffer;
+
+  // For multi-value column only
+  private int _valueBufferOffset;
+  private ByteBuffer _numValuesByteBuffer;
+  private IntBuffer _numValuesIntBuffer;
+
+  // Posting list related buffers
+  // Constructed in first pass
+  private ByteBuffer _postingListLengthByteBuffer;
+  private IntBuffer _postingListLengthIntBuffer;
+  // Constructed in second pass
+  private ByteBuffer _postingListValueByteBuffer;
+  private ByteBuffer _postingListCurrentOffsetByteBuffer;
+
+  public OffHeapBitmapInvertedIndexCreator(File indexDir, FieldSpec fieldSpec, int cardinality, int numDocs,
```

Comment:
We should be using Mmap for temporary buffers, so that we can load/create multiple segments in parallel without worrying about MaxDirectMemory. Does not have to be in this PR, but please add a TODO.

Reply:
Changing it to mmap buffer will have bad performance impact because of the disk IO (this is already causing problem for star-tree generation). We need to discuss this later when we actually need to generate huge size direct buffers.